src/Language/DifferentialDatalog/OVSDB/Compile.hs

```diff
@@ -0,0 +1,127 @@
+    schemas <- mapM (\fname -> do 
+                      content <- readFile fname
+                      case parseSchema content fname of
+                           Left  e    -> errorWithoutStackTrace $ "Failed to parse input file: " ++ e
+                           Right prog -> return prog)
+                    file_names
+    case compileSchemas schemas outputs of
+         Left e    -> errorWithoutStackTrace e
+         Right doc -> return doc
+
+compileSchemas :: (MonadError String me) => [OVSDBSchema] -> [String] -> me Doc
+compileSchemas schemas outputs = do
+    let tables = concatMap renameTables schemas
+    mapM_ (\o -> do let t = find ((==o) . name) tables
+                    when (isNothing t) $ throwError $ "Table " ++ o ++ " not found") outputs
+    uniqNames ("Multiple declarations of table " ++ ) tables
+    rels <- vcat <$> mapM (\t -> if elem (name t) outputs 
+                                    then mkTable False t
+                                    else mkTable True t) tables
+    return $ pp builtinTypes $+$ rels
+
+renameTables :: OVSDBSchema -> [Table]
+renameTables OVSDBSchema{..} =
+    map (\t -> t{tableName = schemaName ++ "_" ++ tableName t}) schemaTables
+
+mkTable :: (MonadError String me) => Bool -> Table -> me Doc
+mkTable isinput t@Table{..} = do
+    let tprops = filter (\case
+                          ColumnsProperty{} -> True
+                          _                 -> False) tableProperties
+    check (not $ null tprops) (pos t) $ "Table " ++ tableName ++ " does not have a \"columns\" property"
+    check (length tprops == 1) (pos t) $ "Table " ++ tableName ++ " has multiple \"columns\" properties"
+    let (ColumnsProperty ovscols) : _ = tprops
+    uniqNames (\col -> "Multiple declarations of column " ++ col ++ " in table " ++ tableName) ovscols
+    columns <- mapM (mkCol isinput tableName) ovscols
+    let cols   = lefts  columns
+        tables = rights columns
+    let uuidcol = "_uuid: uuid"
+    let prefix = if isinput then "input" else empty
+    return $ prefix <+> "relation" <+> pp (name t) <> "("    $$
+             (nest' $ vcat $ punctuate comma (uuidcol:cols)) $$
+             ")"                                             $$
+             vcat tables
+
+mkCol :: (MonadError String me) => Bool -> String -> TableColumn -> me (Either Doc Doc)
+mkCol isinput tname c@TableColumn{..} = do
+    check (columnName /= "_uuid") (pos c) $ "Illegal column name _uuid in table " ++ tname
+    case columnType of
+         ColumnTypeAtomic at  -> (Left . (mkColName columnName <>) . (":" <+>)) <$> mkAtomicType at
+         ColumnTypeComplex ct -> mkComplexType isinput tname c ct
+
+mkColName :: String -> Doc
+mkColName "type" = "_type"
+mkColName "match" = "_match"
+mkColName n      = pp $ map toLower n
+
+mkAtomicType :: (MonadError String me) => AtomicType -> me Doc
+mkAtomicType IntegerType{} = return "integer"
+mkAtomicType (RealType p)  = err p "\"real\" types are not supported"
```

Comment:
Could this be an external type?
Is this needed?

Reply:
Yep, could be `extern`. There is not a single occurrence of this type in OVN, though, so probably not worth it.